src/TCProcessor.cpp

```diff
@@ -0,0 +1,735 @@
+                      << it->readout_end;
+      }
+      it->contributing_tcs.push_back(tc);
+      break;
+    }
+    ++it;
+  }
+  return;
+}
+
+bool
+TCProcessor::check_overlap(const triggeralgs::TriggerCandidate& tc, const PendingTD& pending_td)
+{
+  if ( (m_use_readout_map) && (m_readout_window_map.count(tc.type)) ) {
+    return !(((tc.time_candidate + m_readout_window_map[tc.type].second) < pending_td.readout_start) ||
+             ((tc.time_candidate - m_readout_window_map[tc.type].first > pending_td.readout_end)));
+  } else {
+    return !((tc.time_end < pending_td.readout_start) || (tc.time_start > pending_td.readout_end));
+  }
+}
+
+std::vector<TCProcessor::PendingTD>
+TCProcessor::get_ready_tds(std::vector<PendingTD>& pending_tds)
+{
+  std::vector<PendingTD> return_tds;
+  for (std::vector<PendingTD>::iterator it = pending_tds.begin(); it != pending_tds.end();) {
+    auto timestamp_now =
+      std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now().time_since_epoch())
+        .count();
+    if (timestamp_now >= it->walltime_expiration) {
+      return_tds.push_back(*it);
+      it = pending_tds.erase(it);
+    } else if (check_td_readout_length(*it)) { // Also pass on TDs with (too) long readout window
+      return_tds.push_back(*it);
+      it = pending_tds.erase(it);
+    } else {
+      ++it;
+    }
+  }
+  return return_tds;
+}
+
+int
+TCProcessor::get_earliest_tc_index(const PendingTD& pending_td)
+{
+  int earliest_tc_index = -1;
+  triggeralgs::timestamp_t earliest_tc_time;
+  for (int i = 0; i < static_cast<int>(pending_td.contributing_tcs.size()); i++) {
+    if (earliest_tc_index == -1) {
+      earliest_tc_time = pending_td.contributing_tcs[i].time_candidate;
+      earliest_tc_index = i;
+    } else {
+      if (pending_td.contributing_tcs[i].time_candidate < earliest_tc_time) {
+        earliest_tc_time = pending_td.contributing_tcs[i].time_candidate;
+        earliest_tc_index = i;
+      }
+    }
+  }
+  return earliest_tc_index;
+}
+
+bool TCProcessor::check_td_readout_length(const PendingTD& pending_td)
+{
+  bool td_too_long = false;
+  if (static_cast<int64_t>(pending_td.readout_end - pending_td.readout_start) >= m_td_readout_limit) {
+    td_too_long = true;
+    TLOG_DEBUG(3) << "Too long readout window: " << (pending_td.readout_end - pending_td.readout_start)
+                  << ", sending immediate TD!";
+  }
+  return td_too_long;
+}
+
+void
+TCProcessor::clear_td_vectors()
+{
+  std::lock_guard<std::mutex> lock(m_td_vector_mutex);
+  m_pending_tds.clear();
+}
+bool
+TCProcessor::check_trigger_type_ignore(unsigned int tc_type)
+{
+  bool ignore = false;
+  for (std::vector<unsigned int>::iterator it = m_ignored_tc_types.begin(); it != m_ignored_tc_types.end();) {
+    if (tc_type == *it) {
+      ignore = true;
+      break;
+    }
+    ++it;
+  }
+  return ignore;
+}
+
+std::bitset<16>
```

Comment:
Thinking out loud, all these 16s will need to be changed to 64s when merging v4 into develop.

I think v4 and v5 diverged so much, that maybe there's some motivation to NOT do a standard v4 merge into develop, and instead just re-implement whatever v4 features we had into v5 from scratch. It might be easier...